File: 0010-regular-expression-matching/0010-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        @cache
        def dp(i, j):
            print(i, j)
            if i < 0 and j < 0:
                return True
            if j < 0:
                return False
            if i < 0:
                if p[j] == '*':
                    return dp(i, j - 2)
                return False

            is_possible = False
            if s[i] == p[j] or p[j] == '.':
                is_possible |= dp(i - 1, j - 1)
            if p[j] == '*':
                regex_char = p[j - 1]
                if s[i] == regex_char or regex_char == '.':
                    is_possible |= dp(i - 1, j)
                    is_possible |= dp(i - 1, j - 2)
                is_possible |= dp(i, j - 2)
            return is_possible
        
        return dp(len(s) - 1, len(p) - 1)

File: 0010-regular-expression-matching/0010-regular-expression-matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        
   std::function<bool(int, int)> find_matches = [&](int i, int j) -> bool {
        if (j >= p.size()) return i >= s.size();
        
        bool first_match = (i < s.size()) && 
                          (p[j] == s[i] || p[j] == '.');
        
        if (j < p.size() - 1 && p[j + 1] == '*') {
            return find_matches(i, j + 2) || 
                   (first_match && find_matches(i + 1, j));
        } else {
            return first_match && find_matches(i + 1, j + 1);
        }
    };

    return find_matches(0, 0);


    }
};